application/src/main/java/org/togetherjava/tjbot/logging/discord/DiscordLogForwarder.java

```diff
@@ -0,0 +1,154 @@
+     * {@value MAX_PENDING_LOGS} is exceeded, logs are discarded.
+     * <p>
+     * This method is thread-safe.
+     *
+     * @param event the log to forward
+     */
+    void forwardLogEvent(LogEvent event) {
+        if (pendingLogs.size() >= MAX_PENDING_LOGS) {
+            logger.warn(LogMarkers.NO_DISCORD,
+                    """
+                            Exceeded the max amount of logs that can be buffered. \
+                            Logs are forwarded to Discord slower than they pile up. Discarding the latest log...""");
+            return;
+        }
+
+        LogMessage log = LogMessage.ofEvent(event);
+
+        synchronized (pendingLogsLock) {
+            pendingLogs.add(log);
+        }
+    }
+
+    private void processPendingLogs() {
+        try {
+            // Process batch
+            List<LogMessage> logsToProcess = pollLogsToProcessBatch();
+            if (logsToProcess.isEmpty()) {
+                return;
+            }
+
+            List<WebhookEmbed> logBatch = logsToProcess.stream().map(LogMessage::embed).toList();
+
+            webhookClient.send(logBatch);
+        } catch (Exception e) {
+            logger.warn(LogMarkers.NO_DISCORD,
+                    "Unknown error when forwarding pending logs to Discord.", e);
+        }
+    }
+
+    private List<LogMessage> pollLogsToProcessBatch() {
+        int batchSize = Math.min(pendingLogs.size(), MAX_BATCH_SIZE);
+        synchronized (pendingLogsLock) {
+            return Stream.generate(pendingLogs::remove).limit(batchSize).toList();
+        }
+    }
+
+    private record LogMessage(WebhookEmbed embed,
+            Instant timestamp) implements Comparable<LogMessage> {
+
+        private static LogMessage ofEvent(LogEvent event) {
+            String authorName = event.getLoggerName();
+            String title = event.getLevel().name();
+            int colorDecimal = Objects.requireNonNull(LEVEL_TO_AMBIENT_COLOR.get(event.getLevel()));
+            String description =
+                    abbreviate(event.getMessage().getFormattedMessage(), MAX_EMBED_DESCRIPTION);
+            Instant timestamp = Instant.ofEpochMilli(event.getInstant().getEpochMillisecond());
+
+            WebhookEmbed embed = new WebhookEmbedBuilder()
+                .setAuthor(new WebhookEmbed.EmbedAuthor(authorName, null, null))
+                .setTitle(new WebhookEmbed.EmbedTitle(title, null))
+                .setDescription(description)
+                .setColor(colorDecimal)
+                .setTimestamp(timestamp)
+                .build();
+
+            return new LogMessage(embed, timestamp);
+        }
+
+        private static String abbreviate(String text, int maxLength) {
+            if (text.length() < maxLength) {
+                return text;
+            }
+
+            if (maxLength < 3) {
+                return text.substring(0, maxLength);
+            }
+
+            return text.substring(0, maxLength - 3) + "...";
```

Comment:
I don't understand why the behaviour changes if the maxLength is above 3

Reply:
It had to do with the expected UX for that `...` stuff on low limits. Lets say u want to abbreviate `hello` to `3`. Then you dont want it to become just `...` but `hel`. For `4` u want `h...`.

That way it ensures that the message is `maxLength` long. 🤷

Reply:
👍